services/autoscaling/src/simcore_service_autoscaling/modules/ec2.py

```diff
@@ -111,33 +104,35 @@ async def start_aws_instance(
         instance_type: InstanceTypeType,
         tags: dict[str, str],
         startup_script: str,
-    ) -> EC2InstanceData:
+        number_of_instances: int,
+    ) -> list[EC2InstanceData]:
         with log_context(
             logger,
             logging.INFO,
             msg=f"launching AWS instance {instance_type} with {tags=}",
         ):
             # first check the max amount is not already reached
+            filters: list[FilterTypeDef] = [
+                {"Name": "tag-key", "Values": [tag_key]} for tag_key in tags.keys()
+            ]
+            filters.append(
+                {"Name": "instance-state-name", "Values": ["pending", "running"]}
+            )
             if current_instances := await self.client.describe_instances(
-                Filters=[
-                    {"Name": "tag-key", "Values": [tag_key]} for tag_key in tags.keys()
-                ]
+                Filters=filters
             ):
                 if (
                     len(current_instances.get("Reservations", []))
                     >= instance_settings.EC2_INSTANCES_MAX_INSTANCES
-                ) and all(
-                    _is_ec2_instance_running(instance)
-                    for instance in current_instances["Reservations"]
                 ):
                     raise Ec2TooManyInstancesError(
                         num_instances=instance_settings.EC2_INSTANCES_MAX_INSTANCES
                     )
 
             instances = await self.client.run_instances(
                 ImageId=instance_settings.EC2_INSTANCES_AMI_ID,
-                MinCount=1,
-                MaxCount=1,
+                MinCount=number_of_instances,
+                MaxCount=number_of_instances,
```

Comment:
NOTE: at the moment I do take GPUs in account

To answer your question I check for each pending task:
1. what instance it needs
2. whether it can fit in another instance that is already there
3. and I sort the pending tasks before treating them